### ISSUE TITLE
Invoke-WebRequest : Unable to read data from the transport connection: An existing connection was forcibly closed by 

### ISSUE THREAD

**Issue by lifenglsf** [php/setup-php-sdk#12]
I want to compied memcached,but got `Unable to read data from the transport connection: An existing connection was forcibly closed by `，60jobs total.It's only partially successful.

**Comment by cmb69** [comment]
Can you share a link to the workflow that failed?

**Comment by lifenglsf** [comment]
[actions url](https://github.com/lifenglsf/php_memcached_dll/actions/runs/10710888809/job/29698548008)

**Comment by cmb69** [comment]
You should update to php/setup-php-sdk@v0.9 (older version fetch from a download server which is no longer up to date).

**Comment by lifenglsf** [comment]
ok，i'll try

**Comment by lifenglsf** [comment]
[0.9 failed](https://github.com/lifenglsf/php_memcached_dll/actions/runs/10803567117/workflow)
not sure if it's a github actions issue

**Comment by cmb69** [comment]
I see three potential problems:

1. too many uncached jobs; that means a lot of request to the server, possibly even for the same file; the server might just block you
1. setup_matrix still refers to windows.php.net, and especially this server is (or at least was) configured to block too many requests
1. setup-php-sdk has the dep `zip` configured, but this doesn't exist

I guess only (1) is relevant for the issue at hand. At least for testing purposes, consider to trim down the matrix.

Also note that memcached 3.2.0 [DLLs are available on PECL](https://pecl.php.net/package/memcached/3.2.0/windows), and that DLLs for older version might be rebuilt. If you need these, or need more configurable builds, consider to file a ticket at https://github.com/php/php-windows-builder.

**Comment by nikic** [comment]
I've been seeing these errors in apcu as well, and they persisted after updating to v0.9: https://github.com/krakjoe/apcu/actions/runs/10869373572/job/30160490179 I'm not querying windows.php.net anywhere.

**Comment by nikic** [comment]
Based on the log, it's already failing to download `https://github.com/php/php-sdk-binary-tools/archive/refs/tags/php-sdk-2.3.0.zip` from GitHub, not the actual PHP release. This is pretty weird...

**Comment by nikic** [comment]
Maybe we need to use GITHUB_TOKEN to avoid rate limiting? Something like this?
```ps
$token = $env:GITHUB_TOKEN
$headers = @{Authorization = "token $token"}
Invoke-WebRequest $url -Headers $headers -OutFile $temp
```

**Comment by cmb69** [comment]
> Based on the log, it's already failing to download `https://github.com/php/php-sdk-binary-tools/archive/refs/tags/php-sdk-2.3.0.zip` from GitHub, not the actual PHP release. This is pretty weird...

Ah, indeed! The problem might be that when downloading archives these may need to be created first; if that takes too long (or something bad happens), the connection might be closed. So maybe downloading a release asset is the better approach.

Can you try https://github.com/php/setup-php-sdk/tree/cmb/download-sdk-release?

> Maybe we need to use GITHUB_TOKEN to avoid rate limiting?

Since so far we're not using the REST API, I don't think this will help. Maybe using the REST API and requiring a GITHUB_TOKEN would improve the situation.

**Comment by nikic** [comment]
> > Based on the log, it's already failing to download `https://github.com/php/php-sdk-binary-tools/archive/refs/tags/php-sdk-2.3.0.zip` from GitHub, not the actual PHP release. This is pretty weird...
> 
> Ah, indeed! The problem might be that when downloading archives these may need to be created first; if that takes too long (or something bad happens), the connection might be closed. So maybe downloading a release asset is the better approach.
> 
> Can you try https://github.com/php/setup-php-sdk/tree/cmb/download-sdk-release?

Seems to work. I ran all 14 jobs three times and they all passed.

> > Maybe we need to use GITHUB_TOKEN to avoid rate limiting?
> 
> Since so far we're not using the REST API, I don't think this will help. Maybe using the REST API and requiring a GITHUB_TOKEN would improve the situation.

Hm, I thought that GitHub uses the Authorization header for everything, not just the REST API. (After all, you could also be trying to download a private project tarball...)

**Comment by cmb69** [comment]
> Seems to work. I ran all 14 jobs three times and they all passed.

Good news. Thanks for checking! I've opened #13.

> Hm, I thought that GitHub uses the Authorization header for everything, not just the REST API. (After all, you could also be trying to download a private project tarball...)

I think they also support unauthenticated downloads for public repos. After all, this is just a web request from another GH hosted runner, but could also be a download via curl or a browser from some local machine, etc.

**Comment by cmb69** [comment]
@lifenglsf, could you please try again with `setup-php-sdk@8d9e79bdbce494591e33d542377778613cfca59f` instead of `setup-php-sdk@v0.9`?

**Comment by lifenglsf** [comment]
ok

**Comment by lifenglsf** [comment]
> I see three potential problems:
> 
> 1. too many uncached jobs; that means a lot of request to the server, possibly even for the same file; the server might just block you
> 2. setup_matrix still refers to windows.php.net, and especially this server is (or at least was) configured to block too many requests
> 3. setup-php-sdk has the dep `zip` configured, but this doesn't exist
> 
> I guess only (1) is relevant for the issue at hand. At least for testing purposes, consider to trim down the matrix.
> 
> Also note that memcached 3.2.0 [DLLs are available on PECL](https://pecl.php.net/package/memcached/3.2.0/windows), and that DLLs for older version might be rebuilt. If you need these, or need more configurable builds, consider to file a ticket at https://github.com/php/php-windows-builder.

i'll try later

**Comment by lifenglsf** [comment]
reducing jobs successful

**Comment by cmb69** [comment]
Okay, I'm closing since the issue appears to have been resolved via 8d9e79bdbce494591e33d542377778613cfca59f. d07cd9875fcb041bbfa63c2c592c4e68a137e1ca can be a further improvement.
